conda_smithy/templates/build_utils.py.tmpl

```diff
@@ -0,0 +1,127 @@
+try:
+    from ruamel_yaml import safe_load, safe_dump
+except ImportError:
+    from yaml import safe_load, safe_dump
+import os
+import subprocess
+import json
+import argparse
+
+global_config = json.loads('''
+{{ channels | tojson }}''')
+call = subprocess.check_call
+
+
+def setup_conda_rc(args):
+    specific_config = safe_load(open(args.config_file))
+    if 'channel_sources' in specific_config:
+        # Due to rendering we may have more than one row for channel_sources, if nothing gets zipped with it
+        first_row = specific_config['channel_sources'][0]  # type: str
+        channels = [c.strip() for c in first_row.split(',')]
+    else:
+        channels = global_config['sources']
+
+    call(['conda', 'config', '--remove', 'channels', 'defaults'])
+    for c in reversed(channels):
+        call(['conda', 'config', '--add', 'channels', c])
+
+    call(['conda', 'config', '--set', 'show_channel_urls', 'true'])
+
+
+def upload_package(args):
+    specific_config = safe_load(open(args.config_file))
+    if 'channel_targets' in specific_config:
+        channels = [c.strip().split(' ') for c in specific_config['channel_targets']]
+    else:
+        channels = global_config['targets']
+
+    for owner, channel in channels:
+        call(['{{ upload_script }}', args.recipe_root, owner, '--channel=%s' % channel, '-m', args.config_file])
+
+
+def make_build_number(args):
+    """
+    General logic
+
+        The purpose of this is to ensure that the new compilers have build
+        numbers > 1000 and legacy compilers have a build number < 1000.
+
+        This is done by reading the build_number_decrement which is rendered
+        into all the recipes.
+
+        For linux and osx we want to avoid building for the legacy compilers
+        with build numbers > 1000
+
+    Example matrix
+        - {'compiler_c': 'toolchain_c', 'build_number_decrement': 1000}
+        - {'compiler_c': 'gcc',         'build_number_decrement': 0}
+
+    """
+    specific_config = safe_load(open(args.config_file))
+    build_number_dec = int(specific_config.get("build_number_decrement", [0])[0])
+
+    use_legacy_compilers = False
+    for key in {'c_compiler', 'cxx_compiler', 'fortran_compiler'}:
+        if 'toolchain_' in specific_config.get(key, ''):
+            use_legacy_compilers = True
+            break
+
+    import conda_build.api
+    rendered_recipe = conda_build.api.render(recipe_path=args.recipe_root, variants=specific_config)
+    build_numbers = set()
+    for recipe, _, _ in rendered_recipe:
+        build_numbers.add(recipe.get_value('build/number'))
+    if len(build_numbers) > 1:
+        raise ValueError("More than one build number found, giving up")
+    try:
+        build_number_int = int(build_numbers.pop())
+
+        if build_number_int < 1000:
+            if not use_legacy_compilers:
+                raise ValueError("Only legacy compilers only valid with build numbers < 1000")
```

Comment:
Instead of erroring, can we remove new compiler configuration?

Reply:
This will catch the situation where we want to turn on new compilers but forgot to build number bump.